src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -4304,9 +4327,55 @@ namespace winrt::TerminalApp::implementation
             }
         });
 
+        sui.GithubAuthRequested([weakThis{ get_weak() }](auto&& /*s*/, auto&& /*e*/) {
+            if (auto page{ weakThis.get() })
+            {
+                page->_InitiateGithubAuth();
+            }
+        });
+
         return *settingsContent;
     }
 
+    void TerminalPage::_InitiateGithubAuth()
+    {
+#if defined(WT_BRANDING_DEV)
+        const auto callbackUri = L"ms-terminal-dev://github-auth";
+#elif defined(WT_BRANDING_CANARY)
+        const auto callbackUri = L"ms-terminal-can://github-auth";
+#endif
+
+        const auto randomStateString = _generateRandomString();
+        const auto executeUrl = fmt::format(FMT_COMPILE(L"https://github.com/login/oauth/authorize?client_id=Iv1.b0870d058e4473a1&redirect_uri={}&state={}"), callbackUri, randomStateString);
+        ShellExecute(nullptr, L"open", executeUrl.c_str(), nullptr, nullptr, SW_SHOWNORMAL);
+        Application::Current().as<TerminalApp::App>().Logic().RandomStateString(randomStateString);
+    }
+
+    std::wstring TerminalPage::_generateRandomString()
+    {
+        // Define the character set to use (wide characters)
+        const std::wstring charset = L"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+        const size_t charsetSize = charset.size();
+
+        // Initialize random number generators
+        std::random_device rd;
+        std::mt19937 generator(rd());
+        std::uniform_int_distribution<> lengthDistribution(8, 12);
+        std::uniform_int_distribution<> charDistribution(0, gsl::narrow<int>(charsetSize) - 1);
```

Comment:
When it comes to the random number generators in C++, it's best as if you pretend as if they don't exist, apart from `std::random_device` and the generic helpers (like `uniform_int_distribution`). In this case you're using `mt19937`, the mersenne twister algorithm, which is a very poor RNG for modern standards. It only exists in the STL because it's a timeless classic. It's also not too helpful here, because it's a PRNG and so the entire random string basically consists just out of the single 32-bit initial seed.

`til::gen_random(void* data, uint32_t length)` from `<til/rand.h>` is a wrapper around [RtlGenRandom](https://learn.microsoft.com/en-us/windows/win32/api/ntsecapi/nf-ntsecapi-rtlgenrandom). It's the same API that `std::random_device` uses but more flexible, because it can extract arbitrary amounts of RNG data at once. We can then use wincrypt to immediately turn that into base64.

Something like that:
```cpp
BYTE buffer[16];
til::gen_random(&buffer[0], sizeof(buffer));

wchar_t string[24];
DWORD stringLen = 24;
THROW_IF_WIN32_BOOL_FALSE(CryptBinaryToStringW(&buffer[0], sizeof(buffer), CRYPT_STRING_BASE64URI | CRYPT_STRING_NOCRLF, &string[0], &stringLen));

return winrt::hstring{ &string[0], stringLen };
```

You should also return a HSTRING, since that's what you're using it as right after.